src/compiler/checker.ts

```diff
@@ -4613,11 +4614,13 @@ module ts {
                     for (let i = 0; i < typeParameters.length; i++) {
                         if (target === typeParameters[i]) {
                             let inferences = context.inferences[i];
-                            let candidates = inferiority ?
-                                inferences.secondary || (inferences.secondary = []) :
-                                inferences.primary || (inferences.primary = []);
-                            if (!contains(candidates, source)) candidates.push(source);
-                            break;
+                            if (!inferences.isFixed) {
+                                let candidates = inferiority ?

```

Comment:
Also, the documentation for `primary` and `secondary` are a bit lacking; elaborate on the distinctions and *why* we make these distinctions.